### PR TITLE
Add some post conditions to ReifyQuotes

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala [hunk 1]
```diff
@@ -74,6 +74,21 @@ class ReifyQuotes extends MacroTransformWithImplicits {
 
   override def phaseName: String = "reifyQuotes"
 
+  override def checkPostCondition(tree: Tree)(implicit ctx: Context): Unit = {
+    if (ctx.owner.ownersIterator.exists(_.isTransparentMethod)) {
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Does it make sense to do these tests if `ctx.compilationUnit.containsQuotesOrSplices` is false?

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Yes, some other phase could inset by mistake one of those calls.

File: compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala [hunk 2]
```diff
@@ -74,6 +74,21 @@ class ReifyQuotes extends MacroTransformWithImplicits {
...
+        case tree: Select if tree.symbol == defn.QuotedExpr_~ =>
+          assert(Splicer.canBeSpliced(tree.qualifier))
+        case _ =>
+      }
+    } else if (!tree.isInstanceOf[DefDef]) {
+      assert(tree.symbol != defn.QuotedExpr_~)
+      // assert(tree.symbol != defn.QuotedType_~) // TODO widen ~ type references at stage 0?
+      assert(tree.symbol != defn.QuotedExpr_apply)
+      assert(tree.symbol != defn.QuotedType_apply)
+    }
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Could this be rewritten as?
```scala
tree match {
  case tree: RefTree if !ctx.inTransparentMethod =>
    assert(!tree.symbol.isQuote)
    // assert(!tree.symbol.isSplice) // TODO widen ~ type references at stage 0?
    assert(tree.symbol != defn.QuotedExpr_~)
  case tree: Select if tree.symbol == defn.QuotedExpr_~ =>
    assert(Splicer.canBeSpliced(tree.qualifier))
  case _ =>
}
```

`ctx.inTransparentMethod` is more expensive than a type test

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Changed

File: compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala [hunk 3]
```diff
@@ -74,6 +74,19 @@ class ReifyQuotes extends MacroTransformWithImplicits {
 
   override def phaseName: String = "reifyQuotes"
 
+  override def checkPostCondition(tree: Tree)(implicit ctx: Context): Unit = {
+    val inTransparentMethod = ctx.owner.ownersIterator.exists(_.isTransparentMethod)
+    tree match {
+      case tree: RefTree if !inTransparentMethod =>
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Why not use `ctx.inTransparentMethod` here?

**Pull request merged**